core/Azure.Mcp.Core/src/Services/ProcessExecution/ExternalProcessService.cs

```diff
@@ -26,130 +241,475 @@ public async Task<ProcessResult> ExecuteAsync(
+            return "<unknown>";
         }
     }
 
-    internal record ParseError(
-        int ExitCode,
-        string Error,
-        string Command
-    );
-
-    internal record ParseOutput([property: JsonPropertyName("output")] string Output);
-
-    public void SetEnvironmentVariables(IDictionary<string, string> variables)
+    /// <summary>
+    /// Gets the exit state of a process, defensively handling exceptions that may be thrown by <see cref="Process.HasExited"/>.
+    ///
+    /// See official docs:
+    /// https://learn.microsoft.com/dotnet/api/system.diagnostics.process.hasexited
+    /// </summary>
+    ///
+    /// <param name="process">The process to check.</param>
+    /// <param name="logger">Logger for diagnostic messages.</param>
+    /// <returns>
+    /// An <see cref="ExitState"/> indicating:
+    /// <list type="bullet">
+    ///   <item><description><see cref="ExitStatus.Exited"/> with null exception if the process has exited.</description></item>
+    ///   <item><description><see cref="ExitStatus.NotExited"/> with null exception if the process has not exited.</description></item>
+    ///   <item><description><see cref="ExitStatus.Exited"/> with null exception if <see cref="InvalidOperationException"/> is thrown.</description></item>
+    ///   <item><description>
+    ///     <see cref="ExitStatus.Indeterminate"/> with the exception if <see cref="System.ComponentModel.Win32Exception"/> or <see cref="NotSupportedException"/> is thrown.
+    ///   </description></item>
+    /// </list>
+    /// </returns>
+    public static ExitCheckResult CheckExitState(this Process process, ILogger logger)
     {
-        if (variables == null)
+        try
         {
-            return;
+            return process.HasExited
+                ? new ExitCheckResult(ExitStatus.Exited, CheckException: null)
+                : new ExitCheckResult(ExitStatus.NotExited, CheckException: null);
         }
+        catch (InvalidOperationException e)
+        {
+            // Official docs: "No process is associated with this object." - treat as "already gone".
+            logger.LogDebug(
+            e,
+            "Process.HasExited reported no associated process. Treating as already exited. " +
+            "Process: {ProcessName}, PID: {Pid}", process.SafeName(), process.SafeId());
+            return new ExitCheckResult(ExitStatus.Exited, CheckException: null);
+        }
+        catch (System.ComponentModel.Win32Exception e)
+        {
+            // Failure to read status (access denied, invalid handle, etc.).
+            return new ExitCheckResult(ExitStatus.Indeterminate, CheckException: e);
+        }
+        catch (NotSupportedException e)
+        {
+            // Remote process or unsupported scenario – not a valid case for Azure MCP Server.
+            return new ExitCheckResult(ExitStatus.Indeterminate, CheckException: e);
+        }
```

Comment:
full exception arg names would make this cleaner

```suggestion
        catch (InvalidOperationException checkException)
        {
            // Official docs: "No process is associated with this object." - treat as "already gone".
            logger.LogDebug(
            checkException,
            "Process.HasExited reported no associated process. Treating as already exited. " +
            "Process: {ProcessName}, PID: {Pid}", process.SafeName(), process.SafeId());
            return new ExitCheckResult(ExitStatus.Exited, CheckException: null);
        }
        catch (System.ComponentModel.Win32Exception checkException)
        {
            // Failure to read status (access denied, invalid handle, etc.).
            return new ExitCheckResult(ExitStatus.Indeterminate, checkException);
        }
        catch (NotSupportedException checkException)
        {
            // Remote process or unsupported scenario – not a valid case for Azure MCP Server.
            return new ExitCheckResult(ExitStatus.Indeterminate, checkException);
        }
```